runtime/lua/vim/_extui.lua

```diff
@@ -0,0 +1,110 @@
+local api = vim.api
+local ext = require('vim._extui.shared')
+ext.msg = require('vim._extui.messages')
+ext.cmd = require('vim._extui.cmdline')
+local M = {}
+local lsauid = -1 -- Autocommand id for 'laststatus' WinOpen/Closed handler.
+
+local function ui_callback(event, ...)
+  local handler = ext.msg[event] or ext.cmd[event]
+  if not handler then
+    return
+  end
+  ext.tab_check_wins()
+  handler(...)
+  vim.api.nvim__redraw({
+    flush = true,
+    cursor = handler == ext.cmd[event] and true or nil,
+    win = handler == ext.cmd[event] and ext.wins[ext.tab].cmd or nil,
+  })
+end
+local scheduled_ui_callback = vim.schedule_wrap(ui_callback)
+
+M.enable = function(enable)
+  if not enable then
+    -- Detach and cleanup windows, buffers and autocommands.
+    for _, tab in ipairs(api.nvim_list_tabpages()) do
+      for _, win in pairs(ext.wins[tab] or {}) do
+        api.nvim_win_close(win, true)
+      end
+    end
+    api.nvim_buf_delete(ext.bufs.cmd, {})
+    api.nvim_buf_delete(ext.bufs.msg, {})
+    api.nvim_buf_delete(ext.bufs.more, {})
+    api.nvim_clear_autocmds({ group = ext.augroup })
+    vim.ui_detach(ext.ns)
+    return
+  end
+
+  ext.cmdheight = vim.o.cmdheight
+  vim.ui_attach(ext.ns, { ext_cmdline = true, ext_messages = true }, function(event, ...)
+    if vim.in_fast_event() then
+      scheduled_ui_callback(event, ...)
+    else
+      ui_callback(event, ...)
+    end
+  end)
+  vim.o.cmdheight = ext.cmdheight
+
+  api.nvim_set_hl(ext.ns, 'Normal', { link = 'MsgArea' })
+  api.nvim_set_hl(ext.ns, 'Search', { link = 'MsgArea' })
+  api.nvim_set_hl(ext.ns, 'CurSearch', { link = 'MsgArea' })
+  api.nvim_set_hl(ext.ns, 'IncSearch', { link = 'MsgArea' })
+
+  api.nvim_create_autocmd('VimResized', {
+    group = ext.augroup,
+    callback = function()
+      ext.msg.set_pos(ext.cmdheight, 0)
+    end,
+    desc = 'Manipulate cmdline and message window dimensions after shell resize.',
+  })
+
+  api.nvim_create_autocmd('OptionSet', {
+    group = ext.augroup,
+    pattern = { 'cmdheight', 'laststatus', 'termguicolors' },
+    callback = function(ev)
+      ext.tab_check_wins()
+      if ev.match == 'cmdheight' then
+        -- 'cmdheight' set; (un)hide cmdline window and set its height.
+        ext.cmdheight = vim.v.option_new
+        local cfg = { height = math.max(ext.cmdheight, 1), hide = ext.cmdheight == 0 }
+        pcall(api.nvim_win_set_config, ext.wins[ext.tab].cmd, cfg)
+      elseif ev.match == 'termguicolors' then
+        -- 'termguicolors' toggled; add or remove border and set 'winblend' for message windows.
+        for _, tab in ipairs(api.nvim_list_tabpages()) do
+          local win = ext.wins[tab].msg
+          pcall(api.nvim_win_set_config, win, { border = vim.v.option_new and 'none' or 'single' })
+          pcall(api.nvim_set_option_value, 'winblend', vim.v.option_new and 30 or 0, { win = win })
```

Comment:
AFAICT winblend only has an effect when `Normal` has a guifg/bg color defined. I keep `Normal` cleared in my colorscheme (it just uses default fg/bg colors from the terminal) so these options make the message window have a plain black background with no border. In addition to setting these options when `'termguicolors'` is set I suggest also checking the values of the `Normal` highlight group.

Reply:
Well the idea was to have an invisible message box (`MsgArea` linked to `Normal` by default); it should not be black unless your terminal background is indeed black, or you define `MsgArea` with a black background.

`'winblend'` helps to discern text that runs behind the floating window, but that doesn't work with `'notermguicolors'`, in which case a border is needed. Otherwise I don't think a cleared `Normal` group is relevant; it should still give alpha to text below the float, but maybe I'm misunderstanding.

Reply:
>Clearing Normal should not yield black background unless your terminal background is indeed black, or you define MsgArea with a black background.

That's what I'm saying, this is not the case. The `'winblend'` option creates a black background when `Normal` is cleared. You can try it for yourself and see. I don't know if this is intended behavior or not (the compositor presumably needs to know the default fg/bg color to perform the blending which it doesn't know if `Normal` is cleared).

Reply:
I'm not sure what to try, could you share reproduction steps and or a screenshot of what you're seeing?

Reply:
With this PR display any kind of message (e.g. `Ctrl-G`) with `Normal` cleared and then again with `Normal` defined with a gui fg/bg color.

With Normal cleared, I see this:

![Screenshot from 2025-01-03 09-24-24](https://github.com/user-attachments/assets/4295d2bf-5b44-4223-b10d-1961b909aec6)

With Normal defined with the same fg/bg colors as my terminal, I see this:

![Screenshot from 2025-01-03 09-24-35](https://github.com/user-attachments/assets/1bbd4a4b-e65a-4471-9dd8-bbb1b54ab3c2)

Reply:
Interesting, not sure what causes that; it doesn't reproduce from `build/bin/nvim --clean +"hi clear Normal" +"set tgc" +"lua require('_extui').enable(true)"` so more info is needed.